### PR TITLE
[sgen][DO NOT MERGE] nursery canaries + managed allocators

### PR DESCRIPTION
**Pull request by alexischr**
Do not merge as it contains a commit that turns on canaries always (for running the test suite)

Missing a way to export this option to AOT images.

### CODE REVIEW

File: mono/sgen/sgen-gc.h [hunk 1]
```diff
@@ -1042,15 +1042,30 @@ gboolean nursery_canaries_enabled (void);
...
-					else				\
-						g_warning ("CORRUPT CANARY:\naddr->%p\ntype->%s\nexcepted->'%s'\nfound->'%s'\n", (char*) addr, sgen_client_vtable_get_name (SGEN_LOAD_VTABLE ((addr))), CANARY_STRING, canary_copy); \
+					char *window_start, *window_end; \
+					window_start = (char*)(addr) - 128; \
+					if (!sgen_ptr_in_nursery (window_start)) \
+						window_start = sgen_get_nursery_start (); \
+					window_end = (char*)(addr) + 128; \
+					if (!sgen_ptr_in_nursery (window_end)) \
+						window_end = sgen_get_nursery_end (); \
+					fprintf (stderr, "\nCANARY ERROR - Type:%s Size:%d Address:%p Data:\n", sgen_client_vtable_get_name (SGEN_LOAD_VTABLE ((addr))), size,  (char*) addr); \
```
Comment (kumpera) [inline mono/sgen/sgen-gc.h]: Don't use fprintf and stderr. It won't work on mobile.

File: mono/metadata/sgen-mono.c [hunk 2]
```diff
@@ -1252,16 +1252,16 @@ create_allocator (int atype, gboolean slowpath)
...
-		/* size += ALLOC_ALIGN - 1; */
 		mono_mb_emit_ldloc (mb, size_var);
-		mono_mb_emit_icon (mb, SGEN_ALLOC_ALIGN - 1);
-		mono_mb_emit_byte (mb, CEE_ADD);
-		/* size &= ~(ALLOC_ALIGN - 1); */
-		mono_mb_emit_icon (mb, ~(SGEN_ALLOC_ALIGN - 1));
-		mono_mb_emit_byte (mb, CEE_AND);
-		mono_mb_emit_stloc (mb, size_var);
 	}
+	/* size += ALLOC_ALIGN - 1; */
```
Comment (kumpera) [inline mono/metadata/sgen-mono.c]: What's the perf impact of this change? We're now doing more work on the case of small objects.

**Comment by alexischr** [comment]
build

**Comment by alexischr** [comment]
build

File: mono/metadata/sgen-mono.c (hunk 2 above)
Comment (alexischr) [inline mono/metadata/sgen-mono.c]: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=176
FWIW, another PR with just the removal of zeroing on strings change:
http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=179

I can't really make heads or tails with these, it seems that neither matters enough to drown the noise. Maybe I should run the bench a couple more times.

Comment (schani) [inline mono/metadata/sgen-mono.c]: Yeah, that's pretty much noise.

**Comment by alexischr** [comment]
build

**Comment by alexischr** [comment]
benchmark

**Pull request merged**